src/network/lan.cpp

```diff
@@ -0,0 +1,335 @@
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with Freeminer.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "network/lan.h"
+#include <cstdint>
+#include "convert_json.h"
+#include "socket.h"
+#include "log.h"
+#include "settings.h"
+#include "version.h"
+#include "networkprotocol.h"
+#include "server/serverlist.h"
+#include "debug.h"
+#include "json/json.h"
+#include <mutex>
+#include <shared_mutex>
+#include "porting.h"
+#include "threading/thread.h"
+#include "network/address.h"
+
+//copypaste from ../socket.cpp
+#ifdef _WIN32
+#ifndef WIN32_LEAN_AND_MEAN
+#define WIN32_LEAN_AND_MEAN
+#endif
+// Without this some of the network functions are not found on mingw
+#ifndef _WIN32_WINNT
+#define _WIN32_WINNT 0x0501
+#endif
+#include <windows.h>
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#define LAST_SOCKET_ERR() WSAGetLastError()
+typedef SOCKET socket_t;
+typedef int socklen_t;
+#else
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <fcntl.h>
+#include <netdb.h>
+#include <unistd.h>
+#include <arpa/inet.h>
+
+#ifndef __ANDROID__
+#include <ifaddrs.h>
+#define HAVE_IFADDRS 1
+#endif
+
+#define LAST_SOCKET_ERR() (errno)
+typedef int socket_t;
+#endif
+
+const char* adv_multicast_addr = "224.1.1.1";
+const static unsigned short int adv_port = 29998;
+static std::string ask_str;
+
+bool use_ipv6 = true;
+
+lan_adv::lan_adv() : Thread("lan_adv")
+{
+}
+
+void lan_adv::ask()
+{
+	if (!isRunning()) start();
+
+	if (ask_str.empty()) {
+		Json::Value j;
+		j["cmd"] = "ask";
+		ask_str = fastWriteJson(j);
+	}
+
+	send_string(ask_str);
+}
+
+void lan_adv::send_string(const std::string &str)
+{
+	if (g_settings->getBool("enable_ipv6")) {
+		std::vector<uint32_t> scopes;
+		// todo: windows and android
+
+#if HAVE_IFADDRS
+		struct ifaddrs *ifaddr = nullptr, *ifa = nullptr;
+		if (getifaddrs(&ifaddr) < 0) {
+		} else {
+			for (ifa = ifaddr; ifa; ifa = ifa->ifa_next) {
+				if (!ifa->ifa_addr)
+					continue;
+				if (ifa->ifa_addr->sa_family != AF_INET6)
+					continue;
+
+				auto sa = *((struct sockaddr_in6 *)ifa->ifa_addr);
+			if (sa.sin6_scope_id)
+					scopes.push_back(sa.sin6_scope_id);
+
+				/*errorstream<<"in=" << ifa->ifa_name << " a="<<Address(*((struct sockaddr_in6*)ifa->ifa_addr)).serializeString()<<" ba=" << ifa->ifa_broadaddr <<" sc=" << sa.sin6_scope_id <<" fl=" <<  ifa->ifa_flags
+				//<< " bn=" << Address(*((struct sockaddr_in6*)ifa->ifa_broadaddr)).serializeString()
+				<<"\n"; */
+			}
+		}
+		freeifaddrs(ifaddr);
+#endif
+
+		if (scopes.empty())
+			scopes.push_back(0);
+
+		struct addrinfo hints
+		{
+		};
+
+		hints.ai_family = AF_INET6;
+		hints.ai_socktype = SOCK_DGRAM;
+		hints.ai_flags = AI_V4MAPPED | AI_ADDRCONFIG;
+		struct addrinfo *result = nullptr;
+		if (!getaddrinfo("ff02::1", nullptr, &hints, &result)) {
```

Comment:
The address should be defined at the top of the file like the IPV4 multicast address is.
Additionally this is a link-local address for all nodes, isn't this the same as IPv4 broadcast? Which makes the address choice for the IPv4 version of this code more confusing. I'm not sure implementing IPv6 server discovery is even a good idea (doesn't ipv4 work fine for that?) but if it's to be done wouldn't a custom interface ID make more sense? Or is there some extra complication with that?

Reply:
It does make more sense to have the ipv6 address with the ipv4. As for your other questions, I wouldn't know. @proller, would you have any idea?